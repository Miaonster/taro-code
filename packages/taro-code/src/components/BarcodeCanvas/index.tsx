import React, { CSSProperties, forwardRef, useEffect, useImperativeHandle, useState } from 'react'
import { Canvas } from '@tarojs/components'
import Taro from '@tarojs/taro'
import { waitForElement } from '../../common/canvas/element'
import { drawCanvasBarcode } from '../../common/canvas/barcode'

export interface BarcodeProps {
  className?: string
  text: string
  scale?: number
  width?: number
  height?: number
  style?: CSSProperties
  foregroundColor?: string
  backgroundColor?: string
  rootSelector?: string
}

const BarcodeCanvas = forwardRef<{ canvas?: Taro.Canvas }, BarcodeProps>((props, ref) => {
  const {
    className,
    text = '',
    width = 300,
    height = 60,
    style = {},
    foregroundColor = '#000000',
    backgroundColor = '#FFFFFF',
    rootSelector,
  } = props
  const id = 'taro-code-canvas-barcode'
  const widthString = width != null ? `${width}px` : ''
  const heightString = height != null ? `${height}px` : ''
  const finalStyle = { width: widthString, height: heightString, ...style }
  const [canvasOutput, setCanvasOutput] = useState<Taro.Canvas>()

  const drawOnce = async (): Promise<void> => {
    const canvas = await waitForElement(id, rootSelector)
    setCanvasOutput(canvas)
    if (canvas != null) {
      drawCanvasBarcode(text, {
        backgroundColor,
        canvas,
        foregroundColor,
        padding: 0,
        height,
        width,
      })
    }
  }

  useEffect(() => {
    void drawOnce()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props])

  useImperativeHandle(
    ref,
    () => {
      return { canvas: canvasOutput }
    },
    [canvasOutput],
  )

  return <Canvas id={id} canvasId={id} type='2d' className={className} style={finalStyle} />
})

export default BarcodeCanvas
