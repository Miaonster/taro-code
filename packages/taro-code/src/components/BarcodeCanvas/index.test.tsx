import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import BarcodeCanvas from './index'

describe('QRCodeCanvas test', () => {
  it('hello world', () => {
    const wrapper = shallow(<BarcodeCanvas text='hello world' />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  it('40+ words', () => {
    const wrapper = shallow(<BarcodeCanvas text='12345678901234567890123456789012345678901234567890' />)
    expect(wrapper).toMatchSnapshot()
  })
})
