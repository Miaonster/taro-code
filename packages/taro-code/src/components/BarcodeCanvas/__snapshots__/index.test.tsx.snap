// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarcodeCanvas test renders correctly with 40+ words 1`] = `
<canvas
  canvasId="taro-code-canvas-barcode"
  id="taro-code-canvas-barcode"
  style={
    {
      "height": "60px",
      "width": "300px",
    }
  }
  type="2d"
/>
`;

exports[`BarcodeCanvas test renders correctly with hello world 1`] = `
<canvas
  canvasId="taro-code-canvas-barcode"
  id="taro-code-canvas-barcode"
  style={
    {
      "height": "60px",
      "width": "300px",
    }
  }
  type="2d"
/>
`;
